File: scripts/match-local-images.ts
/**
 * Match local images by filename similarity + optional overrides; copy to public/product-images.
 */
import * as fs from "fs";
import * as path from "path";

const WORKSPACE = process.cwd();
const DEFAULT_DIR = process.env.CTA_LOCAL_IMAGE_DIR ?? WORKSPACE;
const PUBLIC_DIR = path.join(WORKSPACE, "public", "product-images");
const THRESHOLD = Number(process.env.CTA_IMAGE_MATCH_THRESHOLD ?? 0.22);

type Enriched = {
  id: number;
  slug: string;
  name: string;
  category: string;
  shortDescription: string;
  longDescription: string;
  specs: Record<string, string>;
  price: { amount: number; currency: string; display: string } | null;
  images: { src: string; alt: string }[];
  imageStatus: "matched" | "placeholder";
};

function normalize(s: string): string {
  return s
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, " ")
    .trim();
}

function tokens(s: string): Set<string> {
  return new Set(normalize(s).split(" ").filter((w) => w.length > 1));
}

function score(filename: string, product: Enriched): number {
  const stem = path.parse(filename).name.toLowerCase();
  const hay = new Set([
    ...tokens(product.slug.replace(/-/g, " ")),
    ...tokens(product.name),
    ...tokens(product.category),
    ...tokens(stem),
  ]);
  let hit = 0;
  let total = 0;
  for (const t of tokens(stem)) {
    total++;
    if (hay.has(t)) hit++;
  }
  for (const t of tokens(product.slug.replace(/-/g, " "))) {
    total++;
    if (stem.includes(t)) hit += 0.5;
  }
  if (total === 0) return 0;
  return hit / total;
}

function copyFile(src: string, dest: string) {
  fs.mkdirSync(path.dirname(dest), { recursive: true });
  fs.copyFileSync(src, dest);
}

function loadOverrides(): Record<string, string[]> {
  const p = path.join(WORKSPACE, "data", "image-overrides.json");
  if (!fs.existsSync(p)) return {};
  return JSON.parse(fs.readFileSync(p, "utf8"));
}

function imageExtensions(dir: string): string[] {
  return fs
    .readdirSync(dir)
    .filter((f) => /\.(png|jpe?g|webp|gif|svg)$/i.test(f))
    .filter((f) => !f.startsWith("."));
}

function main() {
  const enrichedPath = path.join(WORKSPACE, "data", "products.enriched.json");
  const products = JSON.parse(fs.readFileSync(enrichedPath, "utf8")) as Enriched[];
  const overrides = loadOverrides();
  const candidates = imageExtensions(DEFAULT_DIR).map((f) => path.join(DEFAULT_DIR, f));
  fs.mkdirSync(PUBLIC_DIR, { recursive: true });

  const used = new Set<string>();

  for (const p of products) {
    p.images = [];
    let best: { file: string; s: number } | null = null;
    if (overrides[p.slug]?.length) {
      for (const rel of overrides[p.slug]) {
        const abs = path.isAbsolute(rel) ? rel : path.join(DEFAULT_DIR, rel);
        if (fs.existsSync(abs)) {
          const ext = path.extname(abs);
          const destName = `${p.slug}-1${ext}`;
          const dest = path.join(PUBLIC_DIR, destName);
          copyFile(abs, dest);
          p.images.push({ src: `/product-images/${destName}`, alt: `${p.name} — equipment photo` });
          used.add(abs);
        }
      }
    }
    if (p.images.length === 0) {
      for (const file of candidates) {
        if (used.has(file)) continue;
        const s = score(path.basename(file), p);
        if (!best || s > best.s) best = { file, s };
      }
      if (best && best.s >= THRESHOLD) {
        const ext = path.extname(best.file);
        const destName = `${p.slug}-1${ext}`;
        copyFile(best.file, path.join(PUBLIC_DIR, destName));
        p.images.push({ src: `/product-images/${destName}`, alt: `${p.name} — equipment photo` });
        used.add(best.file);
      }
    }
    p.imageStatus = p.images.length ? "matched" : "placeholder";
  }

  fs.mkdirSync(path.join(WORKSPACE, "data"), { recursive: true });
  fs.writeFileSync(enrichedPath, JSON.stringify(products, null, 2), "utf8");
  const matched = products.filter((x) => x.imageStatus === "matched").length;
  console.log(`Updated ${products.length} products; ${matched} with local images.`);
}

main();

File: scripts/match-local-images.py
#!/usr/bin/env python3
"""Match local images to products by filename fuzzy-scoring.
Copies matched images into public/product-images/ with slug-based names.
Reads data/image-overrides.json for manual assignments."""

import json, os, re, shutil
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data"
DEFAULT_DIR = Path(os.environ.get("CTA_LOCAL_IMAGE_DIR", str(ROOT)))
PUBLIC_IMAGES = ROOT / "public" / "product-images"
THRESHOLD = float(os.environ.get("CTA_IMAGE_MATCH_THRESHOLD", "0.25"))
IMAGE_EXTS = {".png", ".jpg", ".jpeg", ".webp", ".gif", ".svg"}

def normalize(s):
    return re.sub(r"[^a-z0-9]+", " ", s.lower()).strip()

def tokenize(s):
    return set(w for w in normalize(s).split() if len(w) > 1)

def score(filename_stem, product):
    stem_tokens = tokenize(filename_stem)
    product_tokens = (
        tokenize(product["slug"].replace("-", " "))
        | tokenize(product["name"])
        | tokenize(product.get("category", ""))
    )
    if not stem_tokens:
        return 0.0
    hits = sum(1 for t in stem_tokens if t in product_tokens)
    return hits / len(stem_tokens)

def load_overrides():
    p = DATA / "image-overrides.json"
    if p.exists():
        return json.loads(p.read_text(encoding="utf-8"))
    return {}

def list_images(directory):
    try:
        return [
            directory / f
            for f in os.listdir(directory)
            if Path(f).suffix.lower() in IMAGE_EXTS and not f.startswith(".")
        ]
    except FileNotFoundError:
        return []

def main():
    enriched_path = DATA / "products.enriched.json"
    products = json.loads(enriched_path.read_text(encoding="utf-8"))
    overrides = load_overrides()
    candidates = list_images(DEFAULT_DIR)
    PUBLIC_IMAGES.mkdir(parents=True, exist_ok=True)

    used = set()
    stats = {"matched": 0, "override": 0, "placeholder": 0}

    for p in products:
        p["images"] = []
        slug = p["slug"]

        # 1. Manual override wins
        if slug in overrides:
            for rel in overrides[slug]:
                src = Path(rel) if Path(rel).is_absolute() else DEFAULT_DIR / rel
                if src.exists():
                    ext = src.suffix
                    dest_name = f"{slug}-1{ext}"
                    dest = PUBLIC_IMAGES / dest_name
                    shutil.copy2(src, dest)
                    p["images"].append({"src": f"/product-images/{dest_name}", "alt": f"{p['name']} — equipment photo"})
                    used.add(str(src))
            if p["images"]:
                p["imageStatus"] = "matched"
                stats["override"] += 1
                continue

        # 2. Fuzzy filename match
        best = None
        best_score = 0.0
        for candidate in candidates:
            if str(candidate) in used:
                continue
            s = score(candidate.stem, p)
            if s > best_score:
                best_score = s
                best = candidate

        if best and best_score >= THRESHOLD:
            ext = best.suffix
            dest_name = f"{slug}-1{ext}"
            dest = PUBLIC_IMAGES / dest_name
            shutil.copy2(best, dest)
            p["images"].append({"src": f"/product-images/{dest_name}", "alt": f"{p['name']} — equipment photo"})
            used.add(str(best))
            p["imageStatus"] = "matched"
            stats["matched"] += 1
        else:
            p["imageStatus"] = "placeholder"
            stats["placeholder"] += 1

    enriched_path.write_text(json.dumps(products, indent=2), encoding="utf-8")
    print(f"Done. Overrides: {stats['override']}  Fuzzy matched: {stats['matched']}  Placeholders: {stats['placeholder']}")

if __name__ == "__main__":
    main()
